src/avif.c

```diff
@@ -185,14 +185,9 @@ avifResult avifImageCopy(avifImage * dstImage, const avifImage * srcImage, avifP
         if (allocationResult != AVIF_RESULT_OK) {
             return allocationResult;
         }
-        for (int uvPlane = AVIF_CHAN_U; uvPlane <= AVIF_CHAN_V; ++uvPlane) {
-            if (!srcImage->yuvRowBytes[uvPlane]) {
-                // plane is absent. If we're copying from a source without
-                // them, mimic the source image's state by removing our copy.
-                avifFree(dstImage->yuvPlanes[uvPlane]);
-                dstImage->yuvPlanes[uvPlane] = NULL;
-                dstImage->yuvRowBytes[uvPlane] = 0;
-            }
+        if ((dstImage->yuvRowBytes[AVIF_CHAN_U] && !srcImage->yuvRowBytes[AVIF_CHAN_U]) ||
+            (dstImage->yuvRowBytes[AVIF_CHAN_V] && !srcImage->yuvRowBytes[AVIF_CHAN_V])) {
```

Comment:
An alternative check is to check if `srcImage` is non-4:0:0, as the changelog entry says. An advantage is that we can check that before the allocation.

```
diff --git a/src/avif.c b/src/avif.c
index a73c187..222410f 100644
--- a/src/avif.c
+++ b/src/avif.c
@@ -181,19 +181,14 @@ avifResult avifImageCopy(avifImage * dstImage, const avifImage * srcImage, avifP
     avifImageSetMetadataXMP(dstImage, srcImage->xmp.data, srcImage->xmp.size);
 
     if ((planes & AVIF_PLANES_YUV) && srcImage->yuvPlanes[AVIF_CHAN_Y]) {
+        if ((srcImage->yuvFormat != AVIF_PIXEL_FORMAT_YUV400) &&
+            (!srcImage->yuvRowBytes[AVIF_CHAN_U] || !srcImage->yuvRowBytes[AVIF_CHAN_V])) {
+            return AVIF_RESULT_INVALID_ARGUMENT;
+        }
         const avifResult allocationResult = avifImageAllocatePlanes(dstImage, AVIF_PLANES_YUV);
         if (allocationResult != AVIF_RESULT_OK) {
             return allocationResult;
         }
-        for (int uvPlane = AVIF_CHAN_U; uvPlane <= AVIF_CHAN_V; ++uvPlane) {
-            if (!srcImage->yuvRowBytes[uvPlane]) {
-                // plane is absent. If we're copying from a source without
-                // them, mimic the source image's state by removing our copy.
-                avifFree(dstImage->yuvPlanes[uvPlane]);
-                dstImage->yuvPlanes[uvPlane] = NULL;
-                dstImage->yuvRowBytes[uvPlane] = 0;
-            }
-        }
     }
     if ((planes & AVIF_PLANES_A) && srcImage->alphaPlane) {
         const avifResult allocationResult = avifImageAllocatePlanes(dstImage, AVIF_PLANES_A);
```

Note: We can also check for a similar error for 4:0:0:

```
        if ((srcImage->yuvFormat == AVIF_PIXEL_FORMAT_YUV400) &&
            (srcImage->yuvRowBytes[AVIF_CHAN_U] || srcImage->yuvRowBytes[AVIF_CHAN_V])) {
            return AVIF_RESULT_INVALID_ARGUMENT;
        }
```

Reply:
On second thought, I think we should not do the second check I suggested in the previous comment (in the Note). The reason is that for a 4:0:0 image we should avoid accessing its `yuvPlanes[AVIF_CHAN_U]`, `yuvRowBytes[AVIF_CHAN_U]`, etc.